### PR TITLE
REFACTOR-#1531: move `default_to_pandas` into base query_compiler class

### LINKED ISSUE #1531: General implementation of default_to_pandas
Currently base query_compiler contains abstract version of default_to_pandas and its implementation in pandas's query_compiler looks rather general so we may find a way to move this implementation in base class

### PR DESCRIPTION
**Pull request by anmyachev**
Signed-off-by: Anatoly Myachev <anatoly.myachev@intel.com>

## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #1531 
- [ ] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/storage_formats/base/query_compiler.py [hunk 1]
```diff
@@ -132,7 +131,23 @@ def default_to_pandas(self, pandas_op, *args, **kwargs):
...
             The result of the `pandas_op`, converted back to ``BaseQueryCompiler``.
         """
-        pass
+        op_name = getattr(pandas_op, "__name__", str(pandas_op))
+        ErrorMessage.default_to_pandas(op_name)
+        args = (a.to_pandas() if isinstance(a, type(self)) else a for a in args)
+        kwargs = {
+            k: v.to_pandas() if isinstance(v, type(self)) else v
+            for k, v in kwargs.items()
+        }
```
Comment (dchigarev) [inline modin/core/storage_formats/base/query_compiler.py:140]: let's use more comprehensive approach 
```suggestion
        args = try_cast_to_pandas(args)
        kwargs = try_cast_to_pandas(kwargs)
```

https://github.com/modin-project/modin/blob/1a57a542ad0413fae2591d4b4928b652effd01db/modin/utils.py#L480

File: modin/core/storage_formats/base/query_compiler.py [hunk 2]
```diff
@@ -132,7 +131,23 @@ def default_to_pandas(self, pandas_op, *args, **kwargs):
...
+
+        result = pandas_op(self.to_pandas(), *args, **kwargs)
```
Comment (dchigarev) [inline modin/core/storage_formats/base/query_compiler.py:142]:
```suggestion
        result = pandas_op(try_cast_to_pandas(self), *args, **kwargs)
```

**Comment by anmyachev** [comment]
@modin-project/modin-core is it possible to restart the check of docs?

**Comment by anmyachev** [comment]
@dchigarev ready for merge

**Review by dchigarev — APPROVED** [review]

**Comment by YarShev** [comment]
I don't think we need to move `default_to_pandas` into the base QC. The base QC should be generic for all storage formats. I think the method should belong to the PandasQC or DefaultPandasQC (D2PandasQC) as we thought earlier.

**Comment by dchigarev** [comment]
> The base QC should be generic for all storage formats.

`default_to_pandas` implementation is generic for all QCs, so nothing is broken here I guess.

> the method should belong to the PandasQC or DefaultPandasQC

But is there Default2PandasQC though? It seems that currently, the BaseQueryCompiler is the one that handles falling to pandas logic.

**Comment by YarShev** [comment]
So far we don't have a `Default2PandasQC` but we thought about it at some point. I would prefer to have `default_to_pandas` in that class because there might be `<Default>PyarrowQC` that has a `default_to_arrow`.

**Comment by anmyachev** [comment]
> So far we don't have a `Default2PandasQC` but we thought about it at some point. I would prefer to have `default_to_pandas` in that class because there might be `<Default>PyarrowQC` that has a `default_to_arrow`.

I don’t see any options that some of the Query Compilers will not support defaulting to pandas, because for sure not a single execution will be able to implement all the functionality that is in pandas and it will have to redirect some of the functions to pandas.

**Comment by YarShev** [comment]
PyarrowOnEngine, NumpyOnEngine.

**Comment by anmyachev** [comment]
> PyarrowOnEngine, NumpyOnEngine.

`PyarrowOnEngine` - This is just an implementation detail of the pandas interface. Or don't you think so?

`NumpyOnEngine` - Are you talking about the implementation of the NumPy interface? In this case, we will need a separate hierarchy of QueryCompilers for it. Because now this class aimed to work only with Pandas interface: `This class defines common query compilers API, most of the methods
    are already implemented and defaulting to pandas.` That is, it is not basic for **everything**, but only for the pandas.

**Comment by YarShev** [comment]
> > PyarrowOnEngine, NumpyOnEngine.
> 
> `PyarrowOnEngine` - This is just an implementation detail of the pandas interface. Or don't you think so?

I don't think so. Pyarrow is another storage format in Modin, which should follow the logic/idea of the storage format concept.

> > PyarrowOnEngine, NumpyOnEngine.
> 
> `PyarrowOnEngine` - This is just an implementation detail of the pandas interface. Or don't you think so?
> 
> `NumpyOnEngine` - Are you talking about the implementation of the NumPy interface? In this case, we will need a separate hierarchy of QueryCompilers for it. Because now this class aimed to work only with Pandas interface: `This class defines common query compilers API, most of the methods are already implemented and defaulting to pandas.` That is, it is not basic for **everything**, but only for the pandas.

We will likely need to implement a QC for the NumPy storage format. Again, at some point we wanted to split the BaseQC into a D2PQC and a Base/GenericQC but haven't done anything on that front yet.

**Comment by anmyachev** [comment]
> > > PyarrowOnEngine, NumpyOnEngine.
> > 
> > 
> > `PyarrowOnEngine` - This is just an implementation detail of the pandas interface. Or don't you think so?
> 
> I don't think so. Pyarrow is another storage format in Modin, which should follow the logic/idea of the storage format concept.
> 

I see no contradiction to what I wrote.

> > > PyarrowOnEngine, NumpyOnEngine.
> > 
> > 
> > `PyarrowOnEngine` - This is just an implementation detail of the pandas interface. Or don't you think so?
> > `NumpyOnEngine` - Are you talking about the implementation of the NumPy interface? In this case, we will need a separate hierarchy of QueryCompilers for it. Because now this class aimed to work only with Pandas interface: `This class defines common query compilers API, most of the methods are already implemented and defaulting to pandas.` That is, it is not basic for **everything**, but only for the pandas.
> 
> We will likely need to implement a QC for the NumPy storage format. Again, at some point we wanted to split the BaseQC into a D2PQC and a Base/GenericQC but haven't done anything on that front yet.

Likely. I feel like we need a separate issue for this, where we will continue the discussion. Please create it with all your thoughts on this.

**Comment by vnlitvinov** [comment]
What I think **Iaroslav** is trying to express is that `BaseQueryQC` should mostly be an abstract class which has nothing to do with exact syntax of library being implemented.

I think we should treat Modin like "LLVM for ML", so `modin.pandas` is frontend, query compilers are backend (and IR to some extent). So it's basically different abstraction layers, and Modin per se is not designed to be _limited_ to only Pandas API.

So `PandasQueryCompiler` having a `.default_to_pandas()` method is some implementation details which are not applicable for other _input_ APIs.

cc @devin-petersohn

**Comment by anmyachev** [comment]
> What I think **Iaroslav** is trying to express is that `BaseQueryQC` should mostly be an abstract class which has nothing to do with exact syntax of library being implemented.
> 
> I think we should treat Modin like "LLVM for ML", so `modin.pandas` is frontend, query compilers are backend (and IR to some extent). So it's basically different abstraction layers, and Modin per se is not designed to be _limited_ to only Pandas API.
> 
> So `PandasQueryCompiler` having a `.default_to_pandas()` method is some implementation details which are not applicable for other _input_ APIs.
> 
> cc @devin-petersohn

Almost all functions in `BaseQueryCompiler` class are related to pandas. To do this, it will be easier to rename it to something like `BasePandasQueryCompiler` and create a separate class that will be the base for everything.

**Comment by YarShev** [comment]
https://github.com/modin-project/modin/issues/3636

**Pull request merged**